repair/repair.cc

```diff
@@ -1920,20 +1922,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            std::unordered_set<gms::inet_address> dc_nodes;
+            sstring ks_source_dc;
+            if (ignored_in_my_dc < 0) {
+                size_t ignored_in_my_dc = 0;
+                topology.for_each_node([&] (const locator::node* node) {
+                    if (node->dc_rack().dc != source_dc) {
+                        return;
+                    }
+                    if (ignore_nodes.contains(node->endpoint())) {
+                        ++ignored_in_my_dc;
+                    } else {
+                        dc_nodes.insert(node->endpoint());
+                    }
+                });
+            }
+            // Allow repairing in the dc only if there are enough replicas
+            if (!source_dc.empty() && !dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    if (ignored_in_my_dc) {
+                        // Don't take any chances if there are ignored nodes in this dc
```

Comment:
This is going to introduce regression and surprise for users. Imagine a 5 dc cluster. Each has 10 nodes. One node runs rebuild which ignores a single node in the home dc. After this change, it rebuilds from 50 nodes. This will increases the time significantly. We can not silently do this.

Reply:
The problem is that correctness is more important than performance.  It should still be safe to use the source_dc if we can ensure a quorum by subtracting rf - (1 + ignored_in_my_dc).  My hidden assumption was that rf would be 3 and than if one or more nodes in the dc are ignored we can't ensure a quorum anymore.

Reply:
I added finer grain conditions by racking the full racks, with no dead nodes in `source_dc`, so if we lose say 2 nodes in 1 rack out of with rf=3, we can still use `source_dc` safely.